### ISSUE TITLE
UpdateScreen is ignored

### ISSUE THREAD

**Issue by purplemarshmallow** [purplemarshmallow/z64#3]
This is a design flaw in this plugin. The UpdateScreen call from the emulator is ignored in fact. Instead the plugin updates the screen at VI origin change. This has some bad consequences. Character models in Dark Rift disappear because buffers are displayed at the wrong time. Also single buffered games do not work because VI origin never changes (Quake/Quake II underwater, Vigilante8 pause menu, Taz Express pause menu).

The force swap option in this plugin sloves these problems but it's very hacky and it breaks Mario Party. With force swap UpdateScreen is still ignored. Any attempt from me to fix it properly also broke Mario Party. It causes very strange problems broken rendering and flickering.

**Comment by cxd4** [comment]
Perhaps then, remove the early-exit condition altogether, and/or replace with something else?

The documented early-exit condition I had rewritten from angrylion was:

``` c
int rdp_update(void)
{
 /* ... */
    int vactivelines;
    const int x1 = (*GET_GFX_INFO(VI_H_START_REG) >> 16) & 0x03FF;
    const int y1 = (*GET_GFX_INFO(VI_V_START_REG) >> 16) & 0x03FF;
    const int x2 = (*GET_GFX_INFO(VI_H_START_REG) >>  0) & 0x03FF;
    const int y2 = (*GET_GFX_INFO(VI_V_START_REG) >>  0) & 0x03FF;
    const int v_sync = *GET_GFX_INFO(VI_V_SYNC_REG) & 0x000003FF;
    const int is_PAL = (v_sync > 550);
    const int vitype = *GET_GFX_INFO(VI_STATUS_REG) & 0x00000003;

 /* ... */
    vactivelines = v_sync - (is_PAL ? 47 : 37);
    if (vactivelines < 0)
        return 2;
 /* ... */

    if (!(vitype & 2))
        return 3;

 /* ... */
    if ((vitype >> 1) == 0) {
     /* ... Skip FB emulation, but you may or may not want to update the screen. */
        return 0;
    }

 /* Historically you may or may not want to early-exit here as well. */
    if (x2 < x1 || y2 < y1)
        return 0;
```

**Comment by purplemarshmallow** [comment]
z64gl's optimizations are a problem here. The plugin does delayed rendering and this makes things harder. I'm trying to solve this but I have no luck.

**Comment by cxd4** [comment]
Yeah based on what you mentioned I had subconsciously put this investigation off.

I think my plan is that reading from frame buffer pixels in RDRAM (in case there are no RDP-based graphics commands) should be supported by z64gl first, like how most plugins including angrylion have support for it, then it should be more clear to me how to look at this.

Fortunately I can actually RUN the plugin now on 64-bit Linux, so that helps too when I return to this.

**Comment by purplemarshmallow** [comment]
https://github.com/purplemarshmallow/z64/commit/8cf1c0039ead5684ce54ef666085ab9c30fdcc4c
I've separated rendering and VI emulation. I can't see any regressions, I think that's a step in the right direction

The optimizations aren't a problem anymore looks like the problem is z64gl's buggy/incomplete VI emulation. A lot of code in rglDisplayFramebuffers() does not even seem to do anything
https://github.com/purplemarshmallow/z64/blob/master/src/rgl.cpp#L957
